src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/LazyPInvokeHelpers.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Interlocked = System.Threading.Interlocked;
+
+namespace Internal.Runtime.CompilerHelpers
+{
+    /// <summary>
+    /// These methods aid in lazy PInvoke method resolution.
+    /// </summary>
+    internal static class LazyPInvokeHelpers
+    {
+        internal unsafe static IntPtr ResolvePInvoke(MethodFixupCell* pCell)
+        {
+            if (pCell->Target != IntPtr.Zero)
+                return pCell->Target;
+
+            return ResolvePInvokeSlow(pCell);
+        }
+
+        internal unsafe static IntPtr ResolvePInvokeSlow(MethodFixupCell *pCell)
+        {
+            ModuleFixupCell* pModuleCell = pCell->Module;
+            IntPtr hModule = pModuleCell->Handle;
+            if (hModule == IntPtr.Zero)
+            {
+                FixupModuleCell(pModuleCell);
+                hModule = pModuleCell->Handle;
+                if (hModule == IntPtr.Zero)
+                    throw new TypeLoadException();
+            }
+
+            FixupMethodCell(hModule, pCell);
+            IntPtr fnptr = pCell->Target;
+            if (fnptr == IntPtr.Zero)
+                throw new TypeLoadException();
+
+            return fnptr;
+        }
+
+        internal unsafe static void FixupModuleCell(ModuleFixupCell *pCell)
+        {
+            IntPtr hModule = LoadLibraryExW(pCell->ModuleName, IntPtr.Zero, 0);
+            if (hModule != IntPtr.Zero)
+            {
+                var oldValue = Interlocked.CompareExchange(ref pCell->Handle, hModule, IntPtr.Zero);
+                if (oldValue != IntPtr.Zero)
+                {
+                    // Some other thread won the race to fix it up.
+                    FreeLibrary(hModule);
+                }
+            }
+        }
+
+        internal unsafe static void FixupMethodCell(IntPtr hModule, MethodFixupCell *pCell)
+        {
+            pCell->Target = GetProcAddress(hModule, pCell->MethodName);
+        }
+
+        private const string DelayLoadLib = "kernel32";
```

Comment:
And it should be under `src\Common\src\Interop\Windows\mincore` as any other Windows interop.